src/bin/21_descriptor_layout_and_buffer.rs

```diff
@@ -0,0 +1,723 @@
+#[macro_use]
+extern crate vulkano;
+extern crate vulkano_win;
+extern crate winit;
+
+use std::sync::Arc;
+use std::collections::HashSet;
+use std::time::Instant;
+
+use winit::{EventsLoop, WindowBuilder, Window, dpi::LogicalSize, Event, WindowEvent};
+use vulkano_win::VkSurfaceBuild;
+
+use vulkano::instance::{
+    Instance,
+    InstanceExtensions,
+    ApplicationInfo,
+    Version,
+    layers_list,
+    PhysicalDevice,
+};
+use vulkano::instance::debug::{DebugCallback, MessageTypes};
+use vulkano::device::{Device, DeviceExtensions, Queue, Features};
+use vulkano::swapchain::{
+    Surface,
+    Capabilities,
+    ColorSpace,
+    SupportedPresentModes,
+    PresentMode,
+    Swapchain,
+    CompositeAlpha,
+    acquire_next_image,
+    AcquireError,
+};
+use vulkano::format::Format;
+use vulkano::image::{ImageUsage, swapchain::SwapchainImage};
+use vulkano::sync::{self, SharingMode, GpuFuture};
+use vulkano::pipeline::{
+    GraphicsPipeline,
+    GraphicsPipelineAbstract,
+    viewport::Viewport,
+};
+use vulkano::framebuffer::{
+    RenderPassAbstract,
+    Subpass,
+    FramebufferAbstract,
+    Framebuffer,
+};
+use vulkano::command_buffer::{
+    AutoCommandBuffer,
+    AutoCommandBufferBuilder,
+    DynamicState,
+};
+use vulkano::buffer::{
+    immutable::ImmutableBuffer,
+    BufferUsage,
+    BufferAccess,
+    TypedBufferAccess,
+    CpuAccessibleBuffer,
+};
+
+const WIDTH: u32 = 800;
+const HEIGHT: u32 = 600;
+
+const VALIDATION_LAYERS: &[&str] =  &[
+    "VK_LAYER_LUNARG_standard_validation"
+];
+
+/// Required device extensions
+fn device_extensions() -> DeviceExtensions {
+    DeviceExtensions {
+        khr_swapchain: true,
+        .. vulkano::device::DeviceExtensions::none()
+    }
+}
+
+#[cfg(all(debug_assertions))]
+const ENABLE_VALIDATION_LAYERS: bool = true;
+#[cfg(not(debug_assertions))]
+const ENABLE_VALIDATION_LAYERS: bool = false;
+
+struct QueueFamilyIndices {
+    graphics_family: i32,
+    present_family: i32,
+}
+impl QueueFamilyIndices {
+    fn new() -> Self {
+        Self { graphics_family: -1, present_family: -1 }
+    }
+
+    fn is_complete(&self) -> bool {
+        self.graphics_family >= 0 && self.present_family >= 0
+    }
+}
+
+#[derive(Copy, Clone)]
+struct Vertex {
+    pos: [f32; 2],
+    color: [f32; 3],
+}
+impl Vertex {
+    fn new(pos: [f32; 2], color: [f32; 3]) -> Self {
+        Self { pos, color }
+    }
+}
+impl_vertex!(Vertex, pos, color);
+
+#[derive(Copy, Clone)]
+struct UniformBufferObject {
+    model: glm::Mat4,
+    view: glm::Mat4,
+    proj: glm::Mat4,
```

Comment:
There are a bunch of `field is never used` warnings here and in some other places.
I also noticed there are now quite a few `clippy` warnings in all steps - would be also nice to clean them up eventually (or `allow` them).

Reply:
Noted. I'm going to clean everything up today and then test on my mac as I haven't done that yet.